packages/library-base/src/commonMain/kotlin/io/realm/RealmQuery.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.realm
+
+import io.realm.internal.Flowable
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+/**
+ * Queries returning [RealmResults].
+ */
+interface RealmElementQuery<E> : Flowable<RealmResults<E>> {
+    fun find(): RealmResults<E>
+
+    // TODO missing fine-grained notifications
+    override fun asFlow(): Flow<RealmResults<E>>
+}
+
+/**
+ * Queries returning a single object.
+ * NOTE: The interaction with primitive queries might be a bit akward
+ */
+// TODO : query should this be done using the raw query string
+interface RealmSingleQuery<E> : Flowable<E> {
+    fun find(): E?
+    // When fine-grained notifications are merged
+    // fun asFlow(): Flow<ObjectChange<E>>
+}
+
+/**
+ * Queries that return scalar values. We cannot express in the type-system which scalar values
+ * we support, so this checks must be done by the compiler plugin or at runtime.
+ */
+interface RealmScalarQuery<E> : Flowable<E> {
+    fun find(): E?
+
+    // TODO : query - not sure the comment below is possible, ask!
+    // These will require a few hacks as Core changelisteners do not support these currently.
+    // Easy work-around is just calling `Results.<aggregateFunction>()` inside the NotifierThread
+//    fun asFlow(): Flow<E>
+}
+
+/**
+ * TODO : query
+ */
+interface RealmQuery<E> : RealmElementQuery<E> {
+
+    fun query(filter: String, vararg arguments: Any?): RealmQuery<E>
+    fun sort(property: String, sortOrder: Sort = Sort.ASCENDING): RealmQuery<E>
+    fun sort(vararg propertyAndSortOrder: Pair<String, Sort>): RealmQuery<E>
+    fun distinct(property: String): RealmQuery<E>
```

Comment:
If we have SORT and DISTINCT at this level, we should probably also have `limit`? 🤔 

So even though these can be supplied in the query string, it might make sense that we primarily treat `query` as predicates and leave handling the result set (sort, distinct, limit) to typed functions?